### PR TITLE
feat(balancer): support upstream host_header and router preserve_host  config for stream tls proxy

### PR DESCRIPTION
**Pull request by oowl**
### Summary

1. support upstream `host_header` config for stream tls proxy, so that `TLS` upstream sni can be configured by `host_header`.
2. support route `preserve_host` config for stream tls proxy.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

https://konghq.atlassian.net/browse/FTI-5234

### CODE REVIEW

File: kong/runloop/handler.lua [hunk 1]
```diff
@@ -1046,10 +1046,22 @@ return {
       var.upstream_host = upstream_url_t.host
     end,
     after = function(ctx)
+      local upstream_scheme = var.upstream_scheme
+
+      local balancer_data = ctx.balancer_data
+      balancer_data.scheme = upstream_scheme -- COMPAT: pdk
+
+      local upstream_host = var.upstream_host
```
Comment (windmgc) [inline kong/runloop/handler.lua:1054]: Do we also need to set `preserve_host == true` here?

Comment (ms2008) [reply kong/runloop/handler.lua:1054]: Since this only involves the stream module, I don't think it would be useful to add it?

Comment (oowl) [reply kong/runloop/handler.lua:1061]: added

File: kong/runloop/balancer/init.lua [hunk 2]
```diff
@@ -472,7 +473,8 @@ local function set_host_header(balancer_data, upstream_scheme, upstream_host, is
 
     var.upstream_host = new_upstream_host
 
-    if is_balancer_phase then
+    -- stream module does not support ngx.balancer.recreate_request
+    if is_balancer_phase and is_http_module then
```
Comment (StarlightIbuki) [inline kong/runloop/balancer/init.lua:477]: Talked offline with @oowl. He says the scenario using this feature is not affected, and the fix for `recreate_request` will be written later.

File: CHANGELOG.md [hunk 3]
```diff
@@ -38,6 +38,8 @@
 
 - Enable `expressions` and `traditional_compatible` router flavor in stream subsystem.
   [#11071](https://github.com/Kong/kong/pull/11071)
+- Make upstream `host_header` and router `preserve_host` config work in stream tls proxy.
```
Comment (StarlightIbuki) [inline CHANGELOG.md:41]: The naming "`host_header`" is somehow misleading. But introducing a new field is also risky. How about making an aliasing?

File: spec/02-integration/05-proxy/31-stream_tls_spec.lua [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      assert(tcp:connect(helpers.get_proxy_ip(true), 19443))
+      assert(tcp:sslhandshake(nil, "ssl-hello.com", false))
+      assert(tcp:send("get_sni\n"))
+      local body = assert(tcp:receive("*a"))
+      assert.equal("ssl-hello.com\n", body)
+      tcp:close()
+    end)
+  end)
+end
+end
```
Comment (StarlightIbuki) [inline spec/02-integration/05-proxy/31-stream_tls_spec.lua:161]: Newline at EoF.

Comment (ms2008) [inline spec/02-integration/05-proxy/31-stream_tls_spec.lua:161]: nit: new line
```suggestion
end

```

**Review by StarlightIbuki — APPROVED** [review]
LGTM. Except for the naming of the config.

File: kong/runloop/balancer/init.lua (hunk 2 above)
Comment (oowl) [reply kong/runloop/balancer/init.lua:477]: I will do a patch for Openresty after 3.4 released.

File: CHANGELOG.md (hunk 3 above)
Comment (oowl) [reply CHANGELOG.md:41]: How to make an aliasing?

File: kong/runloop/balancer/init.lua (hunk 2 above)
Comment (oowl) [reply kong/runloop/balancer/init.lua]: I have read nginx source code, nginx will compile all proxy var when init upstream connection https://github.com/nginx/nginx/blob/master/src/stream/ngx_stream_proxy_module.c#L1070

File: CHANGELOG.md (hunk 3 above)
Comment (oowl) [reply CHANGELOG.md]: I think we do not need to do this, let these two field in the stream proxy seems Ok, and do not need some downstream product change anything.(Konnect or other)

**Review by gruceo — APPROVED** [review]

**Review by kikito — APPROVED** [review]

**Pull request merged**